OAK-11182: Create and test general mechanism for creating MongoDocumentStore fullGC garbage for testing

shodaaan opened the pull request:
- created GenerateFullGCCommand oak-run class with methods for generating garbage of type GAP_ORPHANS and EMPTY_PROPERTIES. All garbage is generated under a specific main node. Also created a cleanup method for deleting all generated garbage
- added call to GenerateFullGCCommand to AvailableModes with specific oak-run input variable
- created GenerateFullGCCommandTest class with unit test methods for generating and cleaning up garbage
- made VersionGarbageCollector.fullGCMode public, so it can be used Command class

Commit:
OAK-10813 : using projection while listing/removing root temp nodes

Commit:
OAK-10813 : using projection while listing children of an node

Commit:
OAK-10823 : replaced db.collection.remove with db.collection.deleteOne/deleteMany to make it compatible with both mongo & mongosh

reschke committed:
OAK-10824: Release Oak 1.64.0 - Candidate Release Notes

ahanikel committed:
OAK-10771 - Add disk cache size stats and issue warning if evicted segment has zero length (#1427)

* OAK-10771 - Log if evicted segment has length zero.

* OAK-10771 - Add diskCacheSize stats

* OAK-10771 - Record cache size and change

---------

Co-authored-by: Axel Hanikel <ahanikel@adobe.com>

Commit:
OAK-10826 - Misleading warning about missing checkpoint

Commit:
OAK-10823 : reverted un-intended changes from var to let/const

reschke committed:
OAK-10824: Release Oak 1.64.0 - Candidate Release Notes

reschke committed:
[maven-release-plugin] prepare release jackrabbit-oak-1.64.0

reschke committed:
[maven-release-plugin] prepare for next development iteration

nfsantos committed:
OAK-10789 - Log paths used for inclusing/exclusion for Mongo regex filters in job summary (#1478)

fabriziofortino committed:
OAK-10811: (oak-search-elastic) reduce contention in IndexTracker (#1466)

* OAK-10811: (oak-search-elastic) reduce contention in IndexTracker

* OAK-10811: ElasticIndexInfoProvider uses node api to get :status info

* OAK-10811: rolled back unneeded change

t-rana committed:
Issues/OAK-10675 (#1385)

* OAK-10675: add service principal support in oak-blob-cloud-azure

* OAK-10675: use user delegation key signed sas for service principal and add license to new files

* OAK-10675: some self review changes

* OAK-10675: set null blob headers as empty string

* OAK-10675: update pom to resolve package dependency

* OAK-10675: add import packages to resolve bundle

* OAK-10770: embed runtime dependencies of azure identity in oak-segment-azure

* OAK-10770: remove unused imports, refactor dependencies

* OAK-10770: add imports, refactor dependencies

* OAK-10675: fix runtime dependency issues, code review

reschke committed:
OAK-10829: examples: update build-helper-maven-plugin to 3.6.0 (#1483)

mbaedke committed:
OAK-10624: o.a.j.o.namepath.JcrPathParser does not accept some valid local names containing '{' or '}'. (#1285)

Fixed issue and added test case.

Co-authored-by: Julian Reschke <reschke@apache.org>

mbaedke committed:
OAK-6756: Convert oak-auth-external to OSGi R7 annotations (#1371)

Done

---------

Co-authored-by: Julian Sedding <jsedding@apache.org>

t-rana committed:
OAK-10780: add azure access token refresh logic (#1441)

* OAK-10780: add azure access token refresh logic

* OAK-10780: modify log statement

* OAK-10780: modify log statement

* OAK-10780: reduce token refresh interval

reschke committed:
OAK-10830: Upgrade jackson-databind dependency to 2.17.1 (#1486)

reschke committed:
Revert "OAK-10830: Upgrade jackson-databind dependency to 2.17.1 (#1486)"

This reverts commit 4a4f2ed767da749f48df26bd1aea73ee262a02ed.

dulceanu committed:
OAK-10833 - Consistency check reports success for repository with SNFE in checkpoints (#1487)

reschke committed:
Revert "Revert "OAK-10830: Upgrade jackson-databind dependency to 2.17.1 (#1486)""

This reverts commit a9d0c1c81295dfde2e3b1f1bb97af589e5238816.

mbaedke committed:
OAK-10621: o.a.j.o.namepath.JcrPathParser does not accept indexed expanded names (#1488)

* OAK-10621: o.a.j.o.namepath.JcrPathParser does not accept indexed expanded names.

Fixed, added unit tests.

reschke committed:
OAK-10824: Release Oak 1.64.0

nfsantos committed:
OAK-10825 - Indexing job: log [TASK:<PHASE>:FAIL] log message for when an indexing phase fails. (#1481)

reschke committed:
OAK-10838: oak-lucene: maven install fails in oak-lucene due to baseline check (#1491)

mbaedke committed:
OAK-6761: Convert oak-blob-plugins to OSGi R7 annotations (#1415)

done

t-rana committed:
Issues/oak 10781 (#1485)

* OAK-10675: add service principal support in oak-blob-cloud-azure

* OAK-10675: use user delegation key signed sas for service principal and add license to new files

* OAK-10675: some self review changes

* OAK-10675: set null blob headers as empty string

* OAK-10675: update pom to resolve package dependency

* OAK-10675: add import packages to resolve bundle

* OAK-10770: embed runtime dependencies of azure identity in oak-segment-azure

* OAK-10770: remove unused imports, refactor dependencies

* OAK-10770: add imports, refactor dependencies

* OAK-10675: fix runtime dependency issues, code review

* OAK-10781: add access token refresh mechanism

* close executor

* reduce token refresh delauy to 1 minute

* retrigger build

rishabhdaim committed:
OAK-10739 : Provide support for FullGC in Mongo DocumentStore (#1454)

Co-authored-by: Stefan Egli <stefanegli@apache.org>
Co-authored-by: Rishabh Kumar <diam@adobe.com>
Co-authored-by: Jose Cordero <corderob@adobe.com>
Co-authored-by: Daniel Iancu <iancu@adobe.com>

Full details in branch: DetailedGC/OAK-10199

stefan-egli committed:
OAK-10843 : temporarily disable flaky tests

adelina-marian committed:
OAK-10837: Add documentation for UT/IT NodeStore fixtures

stefan-egli committed:
OAK-10844 : removing some test combinations and adding logging to help narrow down where time is spent

dulceanu committed:
OAK-10807 - Add oak-run diff to Azure repositories (#1496)

stefan-egli committed:
OAK-10843 : exclude a flaky test combination

stefan-egli committed:
OAK-10844 : marking a slow test as ignored for now

stefan-egli committed:
OAK-10846 : ignoring testPartialMergeRootCleanup as it seems to leak a thread

fabriziofortino committed:
OAK-10841: log error cause when ES cannot process a query (#1492)

reschke committed:
OAK-10847: query-spi: bump log level to ERROR for use of deprecated slf4j.event.Level method (#1500)

reschke committed:
OAK-10852: oak-store-document: disable unreliable tests in BranchCommitGCTest

stefan-egli committed:
OAK-10853 : temporarily disable flaky test

nfsantos committed:
OAK-10851 -  Log START/END/FAIL messages of indexing job phases inside the same package (#1501)

Joscorbe committed:
OAK-10854: Allow oak.findOne to find long path documents

stefan-egli committed:
OAK-10852 : refinement of 46476bcbfc to only ignore flaky mode

reschke committed:
OAK-10856: oak-run: TextPopulatorTest uses transitive dependency com.beust.jcommander.internal.Maps (#1505)

reschke committed:
OAK-10852: disable BranchCommitGCTest unmergedAddsThenMergedAddThenUnmergedRemovesChildren for FullGCMode.ORPHANS_EMPTYPROPS_UNMERGED_BC (#1506)

stefan-egli committed:
OAK-10852 : excluding another flaky test combination

reschke committed:
OAK-10860: oak-lucene: update gson test dependency and stop using deprecated APIs (#1509)

reschke committed:
OAK-10861: avoid use of Tika WriteOutHandler convenience method removed in Tika 2.x (#1510)

dulceanu committed:
OAK-9455 - Improve oak-run check to allow fast remote consistency checks (#1508)

nfsantos committed:
OAK-10868 - Use a constant to represent empty StringValue, to avoid allocating new objects in this case. (#1512)

nfsantos committed:
OAK-10866 - Avoid creating an intermediate String when appending a string value to a JsopBuilder. (#1511)

nfsantos committed:
OAK-10872 - Replace use of StringBuilder by string concatenation in PathUtils.concat(String, String)

stefan-egli committed:
OAK-10870 : New oak-mongo.js methods for sizes of properties, counts … (#1514)

* OAK-10870 : New oak-mongo.js methods for sizes of properties, counts by clusterId and unmerged branch commits

* OAK-10870 : addressed review comments

ionutzpi committed:
OAK-10803 - Modified DocumentPropertyState.java to compress values ba… (#1473)

Co-authored-by: pirlogea <pirlogea@adobe.com>

nscendoni committed:
OAK-10451 UserPrincipalProvider may cause many conflicts when under load (#1128)

Co-authored-by: anchela <angela@apache.org>
Co-authored-by: Marcel Reutegger <marcel.reutegger@gmail.com>

reschke committed:
OAK-10873: store-spi: remove use of Guava CharMatcher in test case (#1516)

* OAK-10873: store-spi: remove use of Guava CharMatcher in test case

* OAK-10873: store-spi: remove use of Guava CharMatcher in test case - use less expensive detection

* OAK-10873: store-spi: remove use of Guava CharMatcher in test case - off by one

* Revert "OAK-10873: store-spi: remove use of Guava CharMatcher in test case - off by one"

This reverts commit d78fe0af400f23ac4f1885b5cca6635981ebb94b.

anchela committed:
OAK-10451 : UserPrincipalProvider may cause many conflicts when under load (add missing licence header)

nfsantos committed:
OAK-10879 - Indexing job: In the index report, list phase timings in chronological order and not alphabetic order (#1519)

reschke committed:
Revert "OAK-10803 - Modified DocumentPropertyState.java to compress values ba… (#1473)"

This reverts commit 163352368343164671231b1ee8f4db75f3925877.

reschke committed:
OAK-10883: document mk: disable flaky test VersionGarbageCollectorIT.testBundledPropUnmergedBCGC for mode ORPHANS_EMPTYPROPS_UNMERGED_BC (#1525)

shodaaan committed:
- create class FullGCOptions which will hold the different parameters used for running full GC
- added variables to Configuration.java for 2 full GC modes: GAP_ORPHANS and EMPTY_PROPERTIES
- udpated VersionGarbageCollector constructor, as well as DocumentNodeStore Builder / Service and DocumentNodeStore to get values received from Configuration via the FullGCOptions class

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

# Conflicts:
#	RELEASE-NOTES.txt
#	oak-api/pom.xml
#	oak-auth-external/pom.xml
#	oak-auth-ldap/pom.xml
#	oak-authorization-cug/pom.xml
#	oak-authorization-principalbased/pom.xml
#	oak-benchmarks-elastic/pom.xml
#	oak-benchmarks-lucene/pom.xml
#	oak-benchmarks-solr/pom.xml
#	oak-benchmarks/pom.xml
#	oak-blob-cloud-azure/pom.xml
#	oak-blob-cloud-azure/src/main/java/org/apache/jackrabbit/oak/blob/cloud/azure/blobstorage/AzureBlobContainerProvider.java
#	oak-blob-cloud-azure/src/test/java/org/apache/jackrabbit/oak/blob/cloud/azure/blobstorage/AzureDataStoreUtils.java
#	oak-blob-cloud/pom.xml
#	oak-blob-plugins/pom.xml
#	oak-blob/pom.xml
#	oak-commons/pom.xml
#	oak-core-spi/pom.xml
#	oak-core/pom.xml
#	oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/ConsoleIndexingReporter.java
#	oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/UserPrincipalProvider.java
#	oak-core/src/test/java/org/apache/jackrabbit/oak/security/user/CachedPrincipalMembershipReaderTest.java
#	oak-doc-railroad-macro/pom.xml
#	oak-doc/pom.xml
#	oak-examples/pom.xml
#	oak-examples/standalone/pom.xml
#	oak-examples/webapp/pom.xml
#	oak-exercise/pom.xml
#	oak-http/pom.xml
#	oak-it-osgi/pom.xml
#	oak-it/pom.xml
#	oak-jackrabbit-api/pom.xml
#	oak-jcr/pom.xml
#	oak-lucene/pom.xml
#	oak-parent/pom.xml
#	oak-pojosr/pom.xml
#	oak-query-spi/pom.xml
#	oak-run-commons/pom.xml
#	oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java
#	oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedTransformTask.java
#	oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreHelper.java
#	oak-run-elastic/pom.xml
#	oak-run/pom.xml
#	oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
#	oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/document/RevisionsCommandTest.java
#	oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/tika/TextPopulatorTest.java
#	oak-search-elastic/pom.xml
#	oak-search/pom.xml
#	oak-security-spi/pom.xml
#	oak-segment-aws/pom.xml
#	oak-segment-azure/pom.xml
#	oak-segment-azure/src/main/java/org/apache/jackrabbit/oak/segment/azure/tool/AzureCheck.java
#	oak-segment-azure/src/main/java/org/apache/jackrabbit/oak/segment/azure/tool/ToolUtils.java
#	oak-segment-remote/pom.xml
#	oak-segment-tar/pom.xml
#	oak-shaded-guava/pom.xml
#	oak-solr-core/pom.xml
#	oak-solr-osgi/pom.xml
#	oak-store-composite/pom.xml
#	oak-store-document/pom.xml
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCRecommendations.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BranchCommitGCTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceConfigurationTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupportTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java
#	oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/util/UtilsTest.java
#	oak-store-spi/pom.xml
#	oak-upgrade/pom.xml
#	pom.xml

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

shodaaan committed:
Merge remote-tracking branch 'origin/trunk' into trunk

shodaaan committed:
Merge branch 'refs/heads/trunk' into trunk_Horia

# Conflicts:
#	oak-run-commons/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreHelper.java
#	oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

shodaaan committed:
- created class GenerateFullGCCommand with embedded class GenerateFullGCOptions that holds all of the parameters for generating fullGC

shodaaan committed:
- added code for generating EMPTY_PROPERTIES and GAP_ORPHANS garbage, using parameters passed through GenerateFullGCOptions
- made VersionGarbageCollector.fullGCMode public, so it can be used Command class
- added call to GenerateFullGCCommand to AvailableModes with specific oak-run input variable

shodaaan committed:
- added code for running generateGarbage continuously for a fixed number of runs with delay between runs

shodaaan committed:
- created skeleton code for test class for GenerateFullGCCommand

shodaaan committed:
- added code for generating garbage with revision in the past according to options parameters
- fixed wrong option names in GenerateFullGCOptions

shodaaan committed:
- created Utils class CommandTestUtils and moved captureSystemOutput methods from RevisionsCommandTest to it, in order to reuse them in GenerateFullGCCommandTest
- added test method to GenerateFullGCCommandTest, and final string variables for the options

shodaaan committed:
- updated GenerateFullGCCommand to use only MongoDocumentStore commands for batch deletion of nodes
- added cleanup functionality (using "clean" subcommand)
- added an ignored unit test method for testing garbage generation with custom MongoDB
- added unit test methods for cleanup garbage sub command and functionality
- added unit test methods for generating EMPTY_PROPS and GAP_ORPHANS garbage

shodaaan committed:
- removed old code from Utils.java

shodaaan committed:
- removed unused file FullGCOptions.java

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/AvailableModes.java:
```diff
@@ -75,5 +75,6 @@ public final class AvailableModes {
             .put("segment-copy", new SegmentCopyCommand())
             .put("server", new ServerCommand())
             .put("purge-index-versions", new LucenePurgeOldIndexVersionCommand())
+            .put("generateFullGC", new GenerateFullGCCommand())
```
```suggestion
            .put("createGarbage", new CreateGarbageCommand())
```

stefan-egli replied:
+1 for renaming the command

shodaaan replied:
Changed. Thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    private List<String> generateGarbageContinuously(GenerateFullGCOptions options, Closer closer) throws IOException, Exception {
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
```
Could you please close the executor once finished.
See https://github.com/shodaaan/jackrabbit-oak/blob/trunk/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStoreThrottlingMetricsUpdater.java#L58 for reference.

shodaaan replied:
Changed. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+     * Root node for fullGC garbage generation.
+     * Necessary in order to allow cleanup of all generated garbage nodes by simply removing the root node.
+     */
+    public static String FULLGC_GEN_ROOT_PATH = "fullGCGenRoot";
```
Would recommend not to add content to root and use something like "/tmp" instead, so eg "/tmp/fullGcTestGarbage" or something like that.

rishabhdaim replied:
how about we use `garbage` to make it clear than we are creating garbage under this folder.

shodaaan replied:
Made both changes. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+     * which is necessary in order for each garbage generation run to be unique and not overwrite previous ones.
+     * If continuous generation is enabled, the index of the run will also be appended to this path.
+     */
+    public static String FULLGC_GEN_BASE_PATH = "fullGCGenTest_";
```
once you are under `FULLGC_GEN_ROOT_PATH` and its name indicates that it is for testing and is garbage, do we still need all path elements to contain `fullGc`?

shodaaan replied:
Changed node names. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+     * @throws DocumentStoreException
+     */
+    @Nullable
+    public <T extends Document> T getDocument(Collection<T> collection, String key, int maxCacheAge)
```
as this is only used by test code I'd move it over there (and if needed create an accessor to `documentNodeStore` for example)

shodaaan replied:
Done. Thank you for the suggestion.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -168,7 +168,7 @@ public class VersionGarbageCollector {
      * might no longer be supported.
      */
-    enum FullGCMode {
+    public enum FullGCMode {
```
not sure if we want this public? what about adding a helper under `org.apache.jackrabbit.oak.plugins.document` in oak-run instead?

that helper could perhaps also have some `includesEmptyProps` / `includesGapOrphans` etc methods - as fullgc modes is going to be a combination of different modes (and the current check is hard coded to only modes 1 and 2)

stefan-egli commented on oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/document/GenerateFullGCCommandTest.java:
```diff
@@ -0,0 +1,293 @@
+
+        String generateBasePath = generatedBasePaths.get(0);
+        for (int i = 0; i < garbageNodesParentCount; i ++) {
+            NodeDocument missingDocument = cmd.getCommand().getDocument(Collection.NODES, "3:/"
```
would suggest to use `Utils.getIdFromPath` etc rather than hard coding the depth - eg changing the base path to be `/tmp/xy` would now require this depth to be fixed
(applies to a few different occasions)

shodaaan replied:
Changed. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+        //1. Create nodes with properties
+        NodeBuilder rootNode = documentNodeStore.getRoot().builder();
+        NodeBuilder garbageRootNode = rootNode.child(FULLGC_GEN_ROOT_PATH);
+        garbageRootNode.child(generationBasePath).setProperty("jcr:primaryType", "nt:file", NAME);
```
would use `org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE` instead

also, would not use `nt:file` as that requires eg `jcr:content` child etc (which wasn't noticed in the test as the merge happens without commithook - something that's a bit risky, but this is probably an exception as it is primarily a test command - but this aspect isn't ideal and might lead to surprises). we could use `org.apache.jackrabbit.oak.spi.nodetype.NT_OAK_UNSTRUCTURED` instead (as that is unorderable, and unstructured, it would be a good fix as it doesn't have any special requirements otherwise).

shodaaan replied:
Changed. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+                        setProperty("jcr:primaryType", "nt:file", NAME);
+
+                if (fullGCMode == VersionGarbageCollector.FullGCMode.EMPTYPROPS || fullGCMode == VersionGarbageCollector.FullGCMode.GAP_ORPHANS_EMPTYPROPS) {
+                    garbageRootNode.child(generationBasePath).child(FULLGC_GEN_PARENT_NODE_PREFIX + i).child(FULLGC_GEN_NODE_PREFIX + j).
```
what might be interesting is adding different types of "child subtrees". IIUC then the it will delete the nodes at level `FULLGC_GEN_PARENT_NODE_PREFIX` (which then leaves its children gap orphans). currently it only creates a 1-level deep child under that node. what would be interesting is to have different cases:
* (later perhaps) 0-level deep but just split docs (we actually probably will miss that type of garbage, that's going to be a new TODO probably, but this test could trigger it)
* 1-level deep (which you already have)
* 2-level deep
* 3-level deep

reason being that it might cause different behavior in fullGC code later on - and thus having slight variations of that garbage could be useful to have

shodaaan replied:
Implemented, thank you for the suggestion.
Added orphansDepth and orphansLevelGap parameters for running the command, which allow for generating gap orphans up to 15 levels deep and also allows for deleting nodes in their hierarchy up to the specified orphansLevelGap level.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+
+        VersionGarbageCollector.FullGCMode fullGCMode = getFullGCMode(options);
+        if (fullGCMode == VersionGarbageCollector.FullGCMode.NONE) {
+            LOG.error("Invalid garbageType specified. Must be one of the following: 1 - EMPTYPROPS, 2 - GAP_ORPHANS, 3 - GAP_ORPHANS_EMPTYPROPS");
```
assume this log message will be outdated soon with more fullGC modes ready for rollout. eg we will support ALL_ORPHANS at some point etc... Perhaps just mention to use a valid fullGc mode (and perhaps add the classname)?

shodaaan replied:
Changed. Thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+            for (int i = 0; i < options.getGarbageNodesParentCount(); i++) {
+
+                sbNodePath.setLength(0);
+                sbNodePath.append("3:/").append(FULLGC_GEN_ROOT_PATH).append("/").append(generationBasePath).append("/").
```
(same suggestion here to use Utils for path composition and id conversion)

shodaaan replied:
Done. Thank you for the suggestion.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+        }
+    }
+
+    public void execute(String... args) throws Exception {
```
throwing Exception is not nice, can't you narrow it down? Also I think it's actually throwing throwable, in which case you dont need to specify throws at all

rishabhdaim replied:
@dare-gartners we are overriding the `execute` method from `Command` interface here. We have to follow the syntax.

shodaaan replied:
As @rishabhdaim said, we have to follow the Command interface. No change made.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+
+    public void execute(String... args) throws Exception {
+        Closer closer = Closer.create();
+        try {
```
if you use try with resources, you wouldn't need to write finally any more. this will be done automatically

shodaaan replied:
Done, thank you for the suggestion.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+        return generateBasePaths;
+    }
+
+    private List<String> generateGarbageContinuously(GenerateFullGCOptions options, Closer closer) throws IOException, Exception {
```
IoException is a child class of Exception, so you don't need to specify both in the throws. And again I'd suggest not throwing Exception and narrowing down to specific exceptions that the function might throw

shodaaan replied:
Removed IOException, just Exception is thrown now.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    private String generateGarbage(GenerateFullGCOptions options, Closer closer, int runIndex,
+                                  DocumentNodeStoreBuilder<?> builder, long timestamp) throws IOException, Exception {
```
why dont you generate/create timestamp in this method instead of passing it down? It seems to be called fro only 1 another private function that does nothing other then calling this function

shodaaan replied:
The timestamp is necessary for the flow of calling generateGarbage from generateGarbageContinuously, where all of the generated garbage must be under the parent with the initial timestamp.
No change made.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+        NodeBuilder garbageRootNode = rootNode.child(FULLGC_GEN_ROOT_PATH);
+        garbageRootNode.child(generationBasePath).setProperty("jcr:primaryType", "nt:file", NAME);
+
+        int nodesCountUnderParent = options.getCreateGarbageNodesCount() / options.getGarbageNodesParentCount();
```
can options.getGarbageNodesParentCount() be 0 ?

rishabhdaim replied:
Yes, this is entirely possible. Good catch @dare-gartners.

shodaaan replied:
Added check for garbage nodes count and parent nodes count from options. Thank you for the suggestion.

dare-gartners commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+                }
+            }
+        }
+        documentNodeStore.merge(rootNode, EmptyHook.INSTANCE, CommitInfo.EMPTY);
```
these 2 operations you do again on Line 305&306 and Line 321/322. If fullGCMode can only have 1 value, wouldn't it be enough to do it just once at the end of all If:s ? 
`documentNodeStore.merge(rootNode, EmptyHook.INSTANCE, CommitInfo.EMPTY);
        documentNodeStore.runBackgroundOperations();`

shodaaan replied:
The merge and runBackgroundOperations calls are necessary after each batch of operations in order for the changes to be applied to the MongoDB database.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+public class GenerateFullGCCommand implements Command {
+    private static final Logger LOG = LoggerFactory.getLogger(GenerateFullGCCommand.class);
+
+    private static final String USAGE = "generateFullGC {<jdbc-uri> | <mongodb-uri>} [options]";
```
```suggestion
    private static final String USAGE = "createGarbage {<jdbc-uri> | <mongodb-uri>} [options]";
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateFullGCCommand.java:
```diff
@@ -0,0 +1,416 @@
+     * The index of the parent node will be appended to this prefix.
+     */
+    public static String FULLGC_GEN_PARENT_NODE_PREFIX = "fullGCParent_";
+    public static String FULLGC_GEN_NODE_PREFIX = "fullGCNode_";
```
I would avoid using FULLGC in variable names.

shodaaan replied:
Done, thank you for the suggestion.

shodaaan committed:
- modified generated garbage path to be under "tmp" node and renamed generated garbage nodes to shorter names
- made executor a class level variable and added close method for it
- renamed files to include "Garbage"
- other small suggested changes

shodaaan committed:
- replaced FULLGC with GARBAGE in constant names

shodaaan committed:
- created GenerateGarbageHelper class with methods for checking the garbageOptionType parameter passed to the generateGarbage Command
- reverted change in VersionGarbageCollector.java

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,427 @@
+            garbageNodesParentCount = parser
+                    .accepts("garbageNodesParentCount", "total number of parent nodes under which to create garbage nodes").withRequiredArg()
+                    .ofType(Integer.class).defaultsTo(1);
+            garbageType = parser
```
We should allow it to create multiple types of garbage.

@stefan-egli wdyt?

Here is a sample code to enable delimited values in parameters: https://github.com/apache/jackrabbit-oak/blob/trunk/oak-run/src/main/java/org/apache/jackrabbit/oak/run/DataStoreCheckCommand.java#L148-L150

shodaaan replied:
In the interest of keeping the functionality as simple as possible, we can do this in a future update of the GenerateGarbageCommand, if necessary.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,427 @@
+public class GenerateGarbageCommand implements Command, Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(GenerateGarbageCommand.class);
+
+    private final ScheduledExecutorService continuousRunExecutor = Executors.newScheduledThreadPool(1);
```
We need to register this executor with `Closer` to close this.

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,427 @@
+        return generateGarbage(options, closer, runIndex, builder, generationTimestamp);
+    }
+
+    private String generateGarbage(GenerateFullGCOptions options, Closer closer, int runIndex,
```
`closer` is not used in this method.

shodaaan replied:
Fixed, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/GenerateGarbageHelper.java:
```diff
@@ -0,0 +1,19 @@
+        return fullGCMode == 0;
+    }
+
+    public static boolean includesEmptyProps(int fullGCMode) {
```
```suggestion
    public static boolean isEmptyProps(int fullGCMode) {
```

shodaaan replied:
Fixed, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/GenerateGarbageHelper.java:
```diff
@@ -0,0 +1,19 @@
+        return fullGCMode == 1;
+    }
+
+    public static boolean includesGapOrphans(int fullGCMode) {
```
```suggestion
    public static boolean isGapOrphans(int fullGCMode) {
```

shodaaan replied:
Fixed, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,427 @@
+        documentNodeStore.runBackgroundOperations();
+
+        //3.1. Generate garbage nodes - GAP_ORPHANS - remove parent nodes
+        if (GenerateGarbageHelper.includesGapOrphans(options.getGarbageType())) {
```
One problem with this approach is that the more garbage types we add in the future, the more if/else statements we need to add.

This would make the code more complex, less readable, and very hard to understand.

How about we create `enums` for fullGcMode with `createGarbage` method and each mode would create its type of garbage?

This would ensure that for adding a new type of garbage, we don't need to update the base class, add a new implementation of garbage type, and plug in the base class.

cc @stefan-egli @Joscorbe

rishabhdaim replied:
A sample implementation could be like this.

`private interface GarbageType {
        void createGarbage();
    }`

    `private class EmptyProps implements GarbageType {

        @Override
        public void createGarbage() {
            for (int i = 0; i < options.getGarbageNodesParentCount(); i++) {
                for (int j = 0; j < nodesCountUnderParent; j++) {
                    garbageRootNode.child(generationBasePath).child(GEN_PARENT_NODE_PREFIX + i).child(GEN_NODE_PREFIX + j).
                            removeProperty(EMPTY_PROPERTY_NAME);
                }
            }

            documentNodeStore.merge(rootNode, EmptyHook.INSTANCE, CommitInfo.EMPTY);
            documentNodeStore.runBackgroundOperations();
        }
    }`

rishabhdaim replied:
All these implementations could be saved inside a `Map` of `fullGcMode --> Implementation` and we can simply iterate over garbageType and create garbage of different types in one simple loop.

shodaaan replied:
Thank you for the suggestion, this is a good idea for the future, when we will have multiple garbageTypes generated by the GenerateGarbageCommand.
For now, since we only have 2 types, one if/else is short and quite readable.

As suggested, I have completely separated the methods for generating EMPTY_PROPS / GAP_ORPHANS garbage.

Will definitely revisit the rest of the suggestion when more garbageType generation methods are implemented.

shodaaan committed:
- added orphansDepth and orphansLevelGap parameters for running the command, which allow for generating gap orphans up to 15 levels deep and also allows for deleting nodes in their hierarchy up to the specified orphansLevelGap level
- separated the garbage generation code for EMPTY_PROPS / GAP_ORPHANS into 2 separate methods
- adjusted existing unit test methods in GenerateGarbageCommandTest and added new ones for the new level / gap parameters
- other small changes from the PR review

Joscorbe commented on oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/document/GenerateFullGCCommandTest.java:
```diff
@@ -0,0 +1,293 @@
+        Closer cmd1Closer = cmdGenerateGarbage.getCloser();
+
+        // cleanup garbage
+        String mongoDBConnString = "mongodb://cm-p107053-e251100-username:3fUEP9UHPD9yP4H@cmp107053e251100r355e01-shard-00-00.wxmqq.mongodb.net:27017,cmp107053e251100r355e01-shard-00-01.wxmqq.mongodb.net:27017,cmp107053e251100r355e01-shard-00-02.wxmqq.mongodb.net:27017/cm-p107053-e251100-database?ssl=true&authSource=admin&replicaSet=atlas-4x602o-shard-0&retryWrites=true&readPreference=primaryPref ...
```
Check this please!

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/GenerateGarbageHelper.java:
```diff
@@ -0,0 +1,19 @@
+package org.apache.jackrabbit.oak.plugins.document;
```
license header is missing

shodaaan replied:
Fixed, thank you for the catch.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/AvailableModes.java:
```diff
@@ -75,5 +75,6 @@ public final class AvailableModes {
             .put("segment-copy", new SegmentCopyCommand())
             .put("server", new ServerCommand())
             .put("purge-index-versions", new LucenePurgeOldIndexVersionCommand())
+            .put("generateGarbage", new GenerateGarbageCommand())
```
oak-run commands need to be lower case. for readability it might be easier to use something like `"generate-garbage"`?

Or .. on second thought .. what about expressing more explicitly that this new command is meant for testing only by .. eg calling it `generate-test-garbage` ?

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+
+    private final ScheduledExecutorService continuousRunExecutor = Executors.newScheduledThreadPool(1);
+
+    private static final String USAGE = "createGarbage {<jdbc-uri> | <mongodb-uri>} [options]";
```
should use the actual command instead of `createGarbage`.

Also, somewhere, not sure where, mandatory options should be mentioned, such as IIUC `garbageNodesCount` at least.

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+        return documentNodeStore;
+    }
+
+    private static class GenerateFullGCOptions extends Utils.NodeStoreOptions {
```
I think this shouldn't have "FullGC" in the name anymore?

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+    /**
+     * The root node name for fullGC garbage generation, one level under tmp.
+     */
+    public static String GARBAGE_GEN_ROOT_NODE_NAME = "fullGC_GarbageRoot";
```
not sure about the name here, and whether it should have `fullGC` in it...
what about `oak-run-generated-test-garbage` ?

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+     */
+    private void getTreeNodePaths(NodeBuilder rootNode, String basePath,
+                                  List<String> treeNodePaths, int nodeLevel) {
+        String childBasePath = basePath.replaceFirst(nodeLevel + ":/", (nodeLevel + 1) + ":/");
```
This seems like a good candidate to make use of  `org.apache.jackrabbit.oak.plugins.document.util.Utils.getIdFromPath` or `org.apache.jackrabbit.oak.plugins.document.Path.fromString`. Working on the Path object would also allow to verify that anything added to `treeNodePaths` is indeed at least 2 levels deep and is under the expected main generated-test-garbage path.

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+
+        documentNodeStore.getDocumentStore().remove(org.apache.jackrabbit.oak.plugins.document.Collection.NODES, childNodePaths);
+        documentNodeStore.merge(rootBuilder, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+        documentNodeStore.runBackgroundOperations();
```
(same as above, shouldn't be necessary)

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli replied:
any particular reason you would like to keep these ? I think they apply to tests but not in this case.

shodaaan replied:
Fixed.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/GenerateGarbageCommand.java:
```diff
@@ -0,0 +1,553 @@
+        }
+
+        documentNodeStore.merge(rootNode, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+        documentNodeStore.runBackgroundOperations();
```
this shouldn't be necessary - automatically happens in the background and at dispose time.

shodaaan replied:
Changed, thank you for the suggestion.

stefan-egli replied:
same here

shodaaan replied:
Fixed.

stefan-egli reviewed (commented):
Added a few more points.

One general question is : as this command is clearly a test command not to be used against a production environment : how could that be emphasized. For example: we could add an explicit: "are you really sure you want to generate test garbage against ...". That would make it clear that the target repository is going to be written to. Wdyt?

shodaaan committed:
- added checks to cleanup garbage methods to only delete documents under the garbage root path
- added additional unit test for cleanup of empty properties garbage
- used path utility methods where possible
- added some method descriptions where relevant
- renamed constants and updated comments and strings to remove "fullGC" from strings

shodaaan commented:
> Added a few more points.
> 
> One general question is : as this command is clearly a test command not to be used against a production environment : how could that be emphasized. For example: we could add an explicit: "are you really sure you want to generate test garbage against ...". That would make it clear that the target repository is going to be written to. Wdyt?

I resolved all of the points, thank you for the suggestions.

Regarding the explicit check, I don't think this is necessary because oak-run is a tool that by design makes important / deep changes in the MongoDB it is run on, so users should always be careful and know the exact consequences of the commands they run. 
The USAGE text clearly states what the command and associated subcommands do.
Also, since oak-run commands are designed to be used "with intent", and not "by accident", having an additional confirmation step for each garbage generation command run would just annoy the user IMHO. I didn't see this check for any other command in oak-run.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/AvailableModes.java:
```diff
@@ -75,5 +75,6 @@ public final class AvailableModes {
             .put("segment-copy", new SegmentCopyCommand())
             .put("server", new ServerCommand())
             .put("purge-index-versions", new LucenePurgeOldIndexVersionCommand())
+            .put("generate-test-garbage", new GenerateGarbageCommand())
```
I would suggest changing `generate` to `create`.
Generate is used with non-tangible things whereas `create` is used with tangible ones.

`create` seems more accurate grammatically.

shodaaan replied:
Fixed, thank you for the suggestion.

shodaaan committed:
- renamed command to "create-test-garbage"

stefan-egli commented:
> Regarding the explicit check, I don't think this is necessary

Fair enough.

Two unrelated additional comments:
* the DocumentNodeStore is not properly disposed. You can see this by running the oak-run command consecutively : the second start will have to wait 2min for the lease timeout of the first non-gracefully shut down run. We should ensure a graceful disposal.
* what might be useful is to add the base path where the garbage is generated to - such that the user immediately knows (and doesn't eg have to consult the source code)

shodaaan committed:
- renamed "Generate" to "Create" in class names and relevant methods
- added documentNodeStore.dispose() calls to CreateGarbageCommand.java

shodaaan commented:
> > Regarding the explicit check, I don't think this is necessary
> 
> Fair enough.
> 
> Two unrelated additional comments:
> 
> * the DocumentNodeStore is not properly disposed. You can see this by running the [oak-run](https://issues.apache.org/jira/browse/OAK-run) command consecutively : the second start will have to wait 2min for the lease timeout of the first non-gracefully shut down run. We should ensure a graceful disposal.
> * what might be useful is to add the base path where the garbage is generated to - such that the user immediately knows (and doesn't eg have to consult the source code)

I added documentNodeStore.dispose() calls to CreateGarbageCommand. Thank you for the suggestion

The root path where garbage is generated (/tmp/oak-run-generated-test-garbage) is specified in the USAGE string of CreateGarbageCommand, so users will know what the path is where garbage is generated.

shodaaan committed:
- added documentNodeStore.dispose() calls to CreateGarbageCommand.java
- added disabled lease check mode setting to DocumentNodeStoreBuilder in GenerateGarbageCommand
- added some more output messages to make the functionality clearer when running oak-run

shodaaan commented:
> > > Regarding the explicit check, I don't think this is necessary
> > 
> > 
> > Fair enough.
> > Two unrelated additional comments:
> > 
> > * the DocumentNodeStore is not properly disposed. You can see this by running the [oak-run](https://issues.apache.org/jira/browse/OAK-run) command consecutively : the second start will have to wait 2min for the lease timeout of the first non-gracefully shut down run. We should ensure a graceful disposal.
> > * what might be useful is to add the base path where the garbage is generated to - such that the user immediately knows (and doesn't eg have to consult the source code)
> 
> I added documentNodeStore.dispose() calls to CreateGarbageCommand. Thank you for the suggestion
> 
> The root path where garbage is generated (/tmp/[oak-run-generated-test-garbage](https://issues.apache.org/jira/browse/OAK-run-generated-test-garbage)) is specified in the USAGE string of CreateGarbageCommand, so users will know what the path is where garbage is generated.


I added proper disposal of DocumentNodeStore for both createGarbage and cleanupGarbage commands. Tested oak-run with repeated runs and disposal works correctly.

I also added some println messages with the base path of the generated garbage, as well as the paths of the documents deleted by the "clean" sub-command.

rishabhdaim commented on oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/document/CreateGarbageCommandTest.java:
```diff
@@ -0,0 +1,437 @@
+        public CreateGarbageCmd(String... args) {
+            // append the default mongodb connection string if one was not provided
+            if (args[0].startsWith("mongodb://")) {
+                this.args = ImmutableList.copyOf(args);
```
please avoid the usage of guava collections. We have this OAK-10685 epic to remove guava from Oak.

You can use `List.copyOf` with only the caveat that it doesn't allow `null` values.

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/test/java/org/apache/jackrabbit/oak/plugins/document/CreateGarbageCommandTest.java:
```diff
@@ -0,0 +1,437 @@
+            if (args[0].startsWith("mongodb://")) {
+                this.args = ImmutableList.copyOf(args);
+            } else {
+                this.args = ImmutableList.<String>builder().add(MongoUtils.URL)
```
same as above.

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/GenerateGarbageHelper.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Helper class to generate garbage for testing purposes.
+ */
+public class GenerateGarbageHelper {
```
```suggestion
public class CreateGarbageHelper {
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+import static org.apache.jackrabbit.oak.run.Utils.createDocumentMKBuilder;
+
+/**
+ * GenerateGarbageCommand generates garbage nodes in the repository in order to allow for testing fullGC functionality.
```
```suggestion
 * CreateGarbageCommand creates garbage nodes in the repository to allow for testing fullGC functionality.
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+            "create-test-garbage {<jdbc-uri> | <mongodb-uri>} <sub-command> [options] ",
+            "where sub-command is one of: ",
+            "  clean - clean up all generated garbage under the specified root node",
+            "  generate - generate garbage nodes in the repository, under the root node tmp/oak-run-generated-test-garbage. ",
```
```suggestion
            "  create - create garbage nodes in the repository, under the root node tmp/oak-run-created-test-garbage. ",
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+            "where sub-command is one of: ",
+            "  clean - clean up all generated garbage under the specified root node",
+            "  generate - generate garbage nodes in the repository, under the root node tmp/oak-run-generated-test-garbage. ",
+            "           Use the --garbageType (required) option to specify the type of garbage to generate",
```
```suggestion
            "           Use the --garbageType (required) option to specify the type of garbage to create",
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+         * Sub-command for generating garbage.
+         * This is the default sub-command to run if none is specified.
+         */
+        static final String CMD_GENERATE = "generate";
```
```suggestion
        static final String CMD_CREATE = "create";
```

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim reviewed (changes requested):
I would replace the usage of `generate` with `create`

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+        }
+    }
+
+    public void execute(String... args) throws Exception {
```
missing `@Override` annotation.

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,574 @@
+    }
+
+    @Override
+    public void close() throws IOException {
```
We can safely remove this since we are registering `continuousRunExecutor` with a `closer` instance along with the `Closeable` interface.

shodaaan replied:
Done, thank you for the suggestion.

shodaaan committed:
- removed ImmutableList usage in CreateGarbageCommand.java
- removed Closeable implementation of CreateGarbageCommand
- some minor renames

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/FullGCHelper.java:
```diff
@@ -70,6 +70,14 @@ public static RevisionVector unmergedBranchCommit(final NodeStore store, final C
+    public static boolean includesGapOrphans(int fullGCMode) {
+        return fullGCMode == 2;
+    }
+
```
these methods are not used. would suggest to either remove them or consolidate them with the same in CreateGarbageHelper

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/CreateGarbageHelper.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class CreateGarbageHelper {
+
+    public static boolean isInvalidGarbageGenerationMode(int fullGCMode) {
+        return fullGCMode == 0;
+    }
+
+    public static boolean isEmptyProps(int fullGCMode) {
+        return fullGCMode == 1;
+    }
+
+    public static boolean isGapOrphans(int fullGCMode) {
+        return fullGCMode == 2;
+    }
```
I think it's ok to have these for now (except they are duplicated by FullGCHelper currently).

I do think though that we could move them to oak-store-document's VersionGarbageCollector - or actually make the FullGCMode in there a bit more than just an enum (or some other reasonable refactoring).

but I think the refactoring can be done later and doesn't have to block this PR

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,568 @@
+     * @param treeNodePaths
+     */
+    private void getTreeNodePaths(NodeBuilder rootNode, String basePath,
+                                  List<String> treeNodePaths) {
```
not sure this is a `get` method? what about a `buildTreeNodePaths`?

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,568 @@
+            });
+
+            execute(closer, args);
+
```
when used in continuous mode (eg by setting `--numberOfRuns 42`) then the code here terminates early and closes the Closer - preventing the continuous mode from practically ever running

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/CreateGarbageCommand.java:
```diff
@@ -0,0 +1,568 @@
+        documentNodeStore = builder.build();
+        closer.register(asCloseable(documentNodeStore));
+
+        NodeBuilder rootNode = documentNodeStore.getRoot().builder();
```
nitpick : it could be somewhat confusing if a `NodeBuilder` is called `somethingNode`. Usually we call it `somethingBuilder` or `something` (as `somethingNode` is typically used for a `NodeState`)
(elsewhere I see you're also using the classic terminology, so this applies only to part of the code)

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -168,7 +168,7 @@ public class VersionGarbageCollector {
      * might no longer be supported.
      */
-    enum FullGCMode {
+     enum FullGCMode {
```
nitpick, detail, but I would remove this, it is an unnecessary change

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli reviewed (commented):
Had troubles getting the continuous mode running and I believe it's broken because of the new closer logic closing it early.

(other than that added a few detail remarks as well)

shodaaan committed:
- modified resource disposal via closer.close for CreateGarbageCommand.java in order to cover all cases, including continuous garbage generation
- updated resource disposal in unit tests to account for changes
- other minor changes / renames

shodaaan commented:
> Had troubles getting the continuous mode running and I believe it's broken because of the new closer logic closing it early.
> 
> (other than that added a few detail remarks as well)

Tested oak-run with all modes (including continuous generation), everything works fine now.

stefan-egli reviewed (approved):
+1 lgtm, started the jenkins run, would suggest to go ahead with merge unless we find a regression (unlikely), and then we can iterate from there

stefan-egli commented:
> would suggest to go ahead with merge unless we find a regression

**squash merge** obviously!
